Kernel/Interrupts/InterruptManagement.cpp

```diff
@@ -0,0 +1,227 @@
+#include <AK/FixedArray.h>
+#include <Kernel/ACPI/MultiProcessorParser.h>
+#include <Kernel/Arch/i386/CPU.h>
+#include <Kernel/Interrupts/IOAPIC.h>
+#include <Kernel/Interrupts/InterruptManagement.h>
+#include <Kernel/Interrupts/PIC.h>
+#include <Kernel/VM/MemoryManager.h>
+
+#define PCAT_COMPAT_FLAG 0x1
+namespace Kernel {
+static InterruptManagement* s_interrupt_management;
+static bool initialized;
+
+InterruptManagement& InterruptManagement::the()
+{
+    if (!initialized && s_interrupt_management == nullptr)
+        s_interrupt_management = new InterruptManagement(false);
+    return *s_interrupt_management;
+}
+
+AdvancedInterruptManagement& AdvancedInterruptManagement::the()
+{
+    ASSERT(s_interrupt_management != nullptr);
+    return *(AdvancedInterruptManagement*)s_interrupt_management;
+}
+
+void InterruptManagement::initialize()
+{
+    if (!initialized)
+        s_interrupt_management = new InterruptManagement(true);
+    initialized = true;
+}
+
+void InterruptManagement::switch_to_pic_mode()
+{
+    ASSERT(initialized);
+    kprintf("Interrupts: Switch to Legacy PIC mode\n");
+    for (auto* irq_controller : InterruptManagement::the().m_interrupt_controllers) {
+        ASSERT(irq_controller != nullptr);
+        if (irq_controller->type() == (u8)IRQControllerType::i82093AA) {
+            irq_controller->hard_disable();
+            dbgprintf("Interrupts: Detected %s - Disabled\n", irq_controller->get_model().characters());
+        } else {
+            dbgprintf("Interrupts: Detected %s\n", irq_controller->get_model().characters());
+        }
+    }
+}
+void InterruptManagement::switch_to_ioapic_mode()
+{
+    ASSERT(initialized);
+    kprintf("Interrupts: Switch to IOAPIC mode\n");
+    if (InterruptManagement::the().m_interrupt_controllers.size() == 1) {
+        if (InterruptManagement::the().get_interrupt_controller(0).type() == (u8)IRQControllerType::i8259) {
+            kprintf("Interrupts: NO IOAPIC detected, Reverting to PIC mode.\n");
+            return;
+        }
+    }
+    for (auto* irq_controller : InterruptManagement::the().m_interrupt_controllers) {
+        ASSERT(irq_controller != nullptr);
+        if (irq_controller->type() == (u8)IRQControllerType::i8259) {
+            irq_controller->hard_disable();
+            dbgprintf("Interrupts: Detected %s - Disabled\n", irq_controller->get_model().characters());
+        } else {
+            dbgprintf("Interrupts: Detected %s\n", irq_controller->get_model().characters());
+        }
+    }
+}
+
+InterruptManagement::InterruptManagement(bool create_default_controller)
+{
+    m_interrupt_controllers.resize(1);
```

Comment:
This is already constructed with size 1 per the initializer at the declaration. No need to resize it again.